package NYT.NChunkClient.NProto;

import "yt_proto/yt/core/misc/proto/guid.proto";
import "yt_proto/yt/core/misc/proto/error.proto";
import "yt_proto/yt/client/node_tracker_client/proto/node_directory.proto";
import "yt/ytlib/chunk_client/proto/session_id.proto";
import "yt/ytlib/chunk_client/proto/chunk_info.proto";
import "yt_proto/yt/client/chunk_client/proto/chunk_meta.proto";
import "yt_proto/yt/client/chunk_client/proto/data_statistics.proto";
import "yt_proto/yt/client/chunk_client/proto/chunk_spec.proto";
import "yt_proto/yt/client/chunk_client/proto/confirm_chunk_replica_info.proto";

////////////////////////////////////////////////////////////////////////////////

message TReqLocateChunks
{
    // Each subrequest is chunk id.
    repeated NYT.NProto.TGuid subrequests = 1;

    optional int32 address_type = 12; // EAddressType
}

message TRspLocateChunks
{
    message TSubresponse
    {
        optional bool missing = 1 [default = false];
        repeated fixed64 replicas = 4;
        optional int32 erasure_codec = 3;

        reserved 2;
    }

    required NNodeTrackerClient.NProto.TNodeDirectory node_directory = 1;
    repeated TSubresponse subresponses = 2;
    optional uint64 revision = 3; // TRevision [default = NullRevision]
}

////////////////////////////////////////////////////////////////////////////////

message TReqLocateDynamicStores
{
    // Each subrequest is dynamic store id.
    repeated NYT.NProto.TGuid subrequests = 1;

    optional int32 address_type = 2;

    // When batched, the union of extension tags of all subrequests is taken.
    optional bool fetch_all_meta_extensions = 3 [default = false];
    repeated int32 extension_tags = 4;
}

message TRspLocateDynamicStores
{
    message TSubresponse
    {
        optional bool missing = 1 [default = false];
        optional NYT.NChunkClient.NProto.TChunkSpec chunk_spec = 2;
    }
    required NNodeTrackerClient.NProto.TNodeDirectory node_directory = 1;
    repeated TSubresponse subresponses = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TReqTouchChunks
{
    // Each subrequest is chunk id.
    repeated NYT.NProto.TGuid subrequests = 1;
}

message TRspTouchChunks
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqAllocateWriteTargets
{
    message TSubrequest
    {
        required NProto.TSessionId session_id = 1;

        // Node addresses (with port number!) that are not allowed to be allocated for this chunk.
        // This does not prevent further balancing to move chunk there.
        repeated string forbidden_addresses = 2;

        // Node addresses (with port number!) that were already allocated for this chunk.
        repeated string allocated_addresses = 7;

        // Host name (without port number!) to be preferred for upload.
        // The client typically puts the name of the local host here to
        // facilitate local uploads.
        optional string preferred_host_name = 3;

        // Desired number of nodes to allocate.
        required int32 desired_target_count = 4;

        // Minimum acceptable number of nodes to allocate.
        required int32 min_target_count = 5;

        // An override for chunk replication factor;
        // used when one wants to upload fewer replicas but still guarantee placement safety.
        optional int32 replication_factor_override = 6;
    }

    repeated TSubrequest subrequests = 1;
}

message TRspAllocateWriteTargets
{
    required NNodeTrackerClient.NProto.TNodeDirectory node_directory = 1;

    message TSubresponse
    {
        reserved 1;
        optional NYT.NProto.TError error = 2;
        repeated fixed64 replicas = 3;
    }

    repeated TSubresponse subresponses = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TChunkExportData
{
    required NYT.NProto.TGuid id = 1;
    required uint32 destination_cell_tag = 2;
}

message TChunkImportData
{
    required NYT.NProto.TGuid id = 1;
    required NYT.NChunkClient.NProto.TChunkInfo info = 2;
    required NYT.NChunkClient.NProto.TChunkMeta meta = 3;
    required int32 erasure_codec = 4; // NErasure::ECodec
    optional NYT.NProto.TGuid chunk_schema_id = 5;
}

message TReqExportChunks
{
    required NYT.NProto.TGuid transaction_id = 1;
    repeated TChunkExportData chunks = 2;
}

message TRspExportChunks
{
    repeated TChunkImportData chunks = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqImportChunks
{
    required NYT.NProto.TGuid transaction_id = 1;
    repeated TChunkImportData chunks = 2;
}

message TRspImportChunks
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqGetChunkOwningNodes
{
    required NYT.NProto.TGuid chunk_id = 1;
}

message TRspGetChunkOwningNodes
{
    message TOwningNode
    {
        required NYT.NProto.TGuid node_id = 1;
        optional NYT.NProto.TGuid transaction_id = 2;
    }

    repeated TOwningNode nodes = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TChunkSealInfo
{
    optional int64 first_overlayed_row_index = 1;
    required int64 row_count = 2;
    required int64 uncompressed_data_size = 3;
    required int64 compressed_data_size = 4;

    reserved 5;
}

////////////////////////////////////////////////////////////////////////////////

message TReqCreateChunk
{
    // The id of the transaction where the chunk should be staged.
    required NYT.NProto.TGuid transaction_id = 1;

    // The name of the account paying for this chunk.
    required string account = 2;

    // Type actual type of chunk to create (e.g. regular, erasure, journal etc).
    required int32 type = 3; // NObjectClient::EObjectType

    // Desired number of copies (including those created by background replication).
    optional int32 replication_factor = 4 [default = 1];

    // Minimum number of replicas to read during seal. Journal chunks only.
    optional int32 read_quorum = 5 [default = 0];

    // Minimum number of replicas to flush during write. Journal chunks only.
    optional int32 write_quorum = 6 [default = 0];

    optional int32 erasure_codec = 7 [default = 0]; // NErasure::ECodec

    // Can this chunk be balanced to other nodes?
    optional bool movable = 8 [default = true];

    // Should this chunk be marked as vital?
    optional bool vital = 9 [default = false];

    // If given, the new chunk is immediately attached to this chunk list.
    optional NYT.NProto.TGuid chunk_list_id = 10;

    // Primary medium.
    required string medium_name = 11;

    // Validate that there is enough disk space available.
    optional bool validate_resource_usage_increase = 12 [default = true];

    // If true then the writer will be appending this chunk to an overlayed journal.
    optional bool overlayed = 13 [default = false];

    // If given, forces master to collocate chunks with coinciding hashes.
    // NB: Zeroes are also treated as nulls.
    optional uint64 consistent_replica_placement_hash = 14;

    // Writer guarantees that the difference between length of any two replicas
    // will not exceed this value. Journal chunks only.
    optional int64 replica_lag_limit = 15;

    // The id of chunk being created. Set only in Sequoia commits.
    optional NYT.NProto.TGuid chunk_id = 16;
}

message TRspCreateChunk
{
    optional NYT.NProto.TError error = 1;
    optional NYT.NChunkClient.NProto.TSessionId session_id = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TReqConfirmChunk
{
    reserved 2;
    required NYT.NProto.TGuid chunk_id = 1;
    // COMPAT(kvk1920)
    optional bool location_uuids_supported = 7 [default = false];
    repeated TConfirmChunkReplicaInfo replicas = 9;

    required TChunkMeta chunk_meta = 3;
    required TChunkInfo chunk_info = 4;
    optional bool request_statistics = 5 [default = false];
    optional NYT.NProto.TGuid schema_id = 8;

    reserved 6;
}

message TRspConfirmChunk
{
    optional NYT.NProto.TError error = 1;
    optional TDataStatistics statistics = 2;

    // COMPAT(akozhikhov): Default is set to 1 for compatibility purposes.
    optional uint64 revision = 3 [default = 1];
}

////////////////////////////////////////////////////////////////////////////////

message TReqSealChunk
{
    required NYT.NProto.TGuid chunk_id = 1;
    required TChunkSealInfo info = 3;

    reserved 2;
}

message TRspSealChunk
{
    optional NYT.NProto.TError error = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqCreateChunkLists
{
    required NYT.NProto.TGuid transaction_id = 1;
    required int32 count = 2;
}

message TRspCreateChunkLists
{
    optional NYT.NProto.TError error = 1;
    repeated NYT.NProto.TGuid chunk_list_ids = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TReqUnstageChunkTree
{
    required NYT.NProto.TGuid chunk_tree_id = 1;
    required bool recursive = 2;
}

message TRspUnstageChunkTree
{
    optional NYT.NProto.TError error = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqAttachChunkTrees
{
    required NYT.NProto.TGuid parent_id = 1;
    repeated NYT.NProto.TGuid child_ids = 2;
    optional bool request_statistics = 3 [default = false];
    optional NYT.NProto.TGuid transaction_id = 4;
}

message TRspAttachChunkTrees
{
    optional NYT.NProto.TError error = 1;
    optional TDataStatistics statistics = 2;
}

////////////////////////////////////////////////////////////////////////////////

// COMPAT(aleksandra-zh)
message TReqExecuteBatch
{
    repeated TReqCreateChunk create_chunk_subrequests = 1;
    repeated TReqConfirmChunk confirm_chunk_subrequests = 2;
    repeated TReqSealChunk seal_chunk_subrequests = 3;
    repeated TReqCreateChunkLists create_chunk_lists_subrequests = 4;
    repeated TReqUnstageChunkTree unstage_chunk_tree_subrequests = 5;
    repeated TReqAttachChunkTrees attach_chunk_trees_subrequests = 6;

    // COMPAT(shakurov): deprecated, use TMulticellSyncExt instead.
    optional bool suppress_upstream_sync = 7 [default = false];
}

message TRspExecuteBatch
{
    repeated TRspCreateChunk create_chunk_subresponses = 1;
    repeated TRspConfirmChunk confirm_chunk_subresponses = 2;
    repeated TRspSealChunk seal_chunk_subresponses = 3;
    repeated TRspCreateChunkLists create_chunk_lists_subresponses = 4;
    repeated TRspUnstageChunkTree unstage_chunk_tree_subresponses = 5;
    repeated TRspAttachChunkTrees attach_chunk_trees_subresponses = 6;
}

////////////////////////////////////////////////////////////////////////////////
